pkg/asset/store.go

```diff
@@ -11,50 +12,86 @@ import (
 type Store interface {
 	// Fetch retrieves the state of the given asset, generating it and its
 	// dependencies if necessary.
-	Fetch(Asset) error
+	// It returns dirty if the asset or any of its parents is loaded from on-disk files.
+	Fetch(Asset) (dirty bool, err error)
 }
 
 // StoreImpl is the implementation of Store.
 type StoreImpl struct {
-	assets map[reflect.Type]Asset
+	// OnDiskFiles stores the any existing files under the cluster directory,
+	// so that each asset can load and reuse them.
+	OnDiskFiles map[string][]byte
+	assets      map[reflect.Type]Asset
 }
 
 // Fetch retrieves the state of the given asset, generating it and its
 // dependencies if necessary.
-func (s *StoreImpl) Fetch(asset Asset) error {
+func (s *StoreImpl) Fetch(asset Asset) (dirty bool, err error) {
 	return s.fetch(asset, "")
 }
 
-func (s *StoreImpl) fetch(asset Asset, indent string) error {
+func (s *StoreImpl) fetch(asset Asset, indent string) (dirty bool, err error) {
 	logrus.Debugf("%sFetching %s...", indent, asset.Name())
+
+	// Return immediately if the asset is found in the cache,
+	// this is because we are doinga depth-first-search, it's guaranteed
+	// that we always fetch the parent before children, so we don't need
+	// to worry about invalidating anything in the cache.
 	storedAsset, ok := s.assets[reflect.TypeOf(asset)]
 	if ok {
 		logrus.Debugf("%sFound %s...", indent, asset.Name())
 		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset).Elem())
-		return nil
+		return false, nil
 	}
 
 	dependencies := asset.Dependencies()
 	parents := make(Parents, len(dependencies))
 	if len(dependencies) > 0 {
 		logrus.Debugf("%sGenerating dependencies of %s...", indent, asset.Name())
 	}
+
+	var anyDirty bool
 	for _, d := range dependencies {
-		err := s.fetch(d, indent+"  ")
+		dt, err := s.fetch(d, indent+"  ")
 		if err != nil {
-			return errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+			return false, errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+		}
+		if dt {
+			anyDirty = true
 		}
 		parents.Add(d)
 	}
 
-	logrus.Debugf("%sGenerating %s...", indent, asset.Name())
-	err := asset.Generate(parents)
+	// Try to load from on-disk files first.
+	found, err := asset.Load(LoadFileWithPattern(s.OnDiskFiles))
 	if err != nil {
-		return errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+		return false, errors.Wrapf(err, "unexpected error when loading asset %s", asset.Name())
+	}
+
+	switch {
+	case anyDirty && found:
+		return false, fmt.Errorf("cannot resolve, both parent assets and current asset %q are dirty", asset.Name())
+	case anyDirty:
+		dirty = anyDirty
```

Comment:
I would rather see `dirty = true` so that I don't have to think about what the value of `anyDirty` is.